services/venice-controller/src/main/java/com/linkedin/venice/controller/systemstore/SystemStoreHealthCheckService.java

```diff
@@ -0,0 +1,164 @@
+import com.linkedin.venice.utils.RetryUtils;
+import com.linkedin.venice.utils.concurrent.VeniceConcurrentHashMap;
+import java.time.Duration;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is used to schedule periodic check of user system stores in Venice cluster to make sure system stores has
+ * good version and is ingesting correctly. This service is expected to be setup in the leader controller of the cluster,
+ * and it will periodically scan all user system stores and collect bad system stores. Parent controller will hit the
+ * controller endpoint to collect information about bad system stores.
+ */
+public class SystemStoreHealthCheckService extends AbstractVeniceService {
+  public static final Logger LOGGER = LogManager.getLogger(SystemStoreHealthCheckService.class);
+  private final ReadWriteStoreRepository storeRepository;
+  private final PushStatusStoreReader pushStatusStoreReader;
+  private final PushStatusStoreWriter pushStatusStoreWriter;
+  private final MetaStoreReader metaStoreReader;
+  private final MetaStoreWriter metaStoreWriter;
+
+  private final int checkPeriodInSeconds;
+  private final AtomicBoolean isRunning = new AtomicBoolean(false);
+  private ScheduledExecutorService checkServiceExecutor;
+
+  private Set<String> unhealthySystemStoreSet = new HashSet<>();
+
+  public SystemStoreHealthCheckService(
+      ReadWriteStoreRepository storeRepository,
+      MetaStoreReader metaStoreReader,
+      MetaStoreWriter metaStoreWriter,
+      PushStatusStoreReader pushStatusStoreReader,
+      PushStatusStoreWriter pushStatusStoreWriter,
+      int systemStoreCheckPeriodInSeconds) {
+    this.storeRepository = storeRepository;
+    this.metaStoreWriter = metaStoreWriter;
+    this.metaStoreReader = metaStoreReader;
+    this.pushStatusStoreWriter = pushStatusStoreWriter;
+    this.pushStatusStoreReader = pushStatusStoreReader;
+    this.checkPeriodInSeconds = systemStoreCheckPeriodInSeconds;
+  }
+
+  /**
+   * Return unhealthy system store name set. This API is expected to be called by parent controller.
+   */
+  public Set<String> getUnhealthySystemStoreSet() {
+    return unhealthySystemStoreSet;
+  }
+
+  @Override
+  public boolean startInner() {
+    checkServiceExecutor = Executors.newScheduledThreadPool(1);
+    isRunning.set(true);
+    checkServiceExecutor.scheduleWithFixedDelay(
+        new SystemStoreHealthCheckTask(),
+        checkPeriodInSeconds,
+        checkPeriodInSeconds,
+        TimeUnit.SECONDS);
+    return true;
+  }
+
+  @Override
+  public void stopInner() {
+    isRunning.set(false);
+    checkServiceExecutor.shutdownNow();
```

Comment:
Hmm, actually I use this intentionally as in this case (controller is shutting down or it is no longer the leader controller) we should not care ongoing data collection task...

Reply:
If I am not wrong, shutting down a controller and its JVM requires all non-daemon threads to shutdown first https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html. So, in the case of controller shutting down, if this executor is not fully down, it could actually prevent the controller JVM from shutting down.

Reply:
For this, I think the current implementation every step we check the isRunning atomic boolean to make early termination. 

I know for other implementations in Venice, we have such pattern:
executorService.shutdown();
executorService.awaitTermination(x, seconds)
if (!executorService.isTerminated()) {
  executorService.shutdownNow();
} 

So my understanding of this pattern is this is still best effort to wait for current submitted task to complete. If the running task does not have early termination logic, it will have the same effect. 
I don't mind changing to above pattern to make it looks more gracefully, but I feel like it is the same. It is basically up to the task logic, unless there is an API to force cancel the task.

Reply:
Reading the [example code](https://docs.oracle.com/javase%2F7%2Fdocs%2Fapi%2F%2F/java/util/concurrent/ExecutorService.html#awaitTermination(long,%20java.util.concurrent.TimeUnit)) from official Javadoc, `shutdown` call seems not very necessary given that this is a single thread pool with just one ever-running Task. I agree with you that it is not guaranteed to always cancel the task, but that part that I like is that, once the executor can not be terminated within the given timeout period, there can have a log line to let us to know the issue. Maybe, the following lines would be good enough in this case.

```
       pool.shutdownNow(); // Cancel currently executing tasks
       // Wait a while for tasks to respond to being cancelled
       if (!pool.awaitTermination(60, TimeUnit.SECONDS))
           System.err.println("Pool did not terminate");
```